### ISSUE TITLE
Memory Leak using woodpecker with kubernetes 

### ISSUE THREAD

**Issue by lara-clink** [woodpecker-ci/woodpecker#4228]
### Component

agent

### Describe the bug

I’ve been encountering what appears to be a memory leak issue when running Woodpecker CI on a Kubernetes cluster. After running pipelines over time, I noticed that the memory usage of the Woodpecker agents and server steadily increases, eventually leading to performance degradation and, in some cases, the need for manual intervention to prevent the system from becoming unresponsive.

### Steps to reproduce

Deploy Woodpecker CI in a Kubernetes environment.
Run multiple pipelines continuously over an extended period.
Monitor memory usage of the Woodpecker agents and server, I will attach my grafana memory usage graph.
Notice that memory consumption increases over time without being released after pipeline execution.
![image](https://github.com/user-attachments/assets/0c4fdc5c-bb6b-498f-855a-bd81a7d116eb)

### Expected behavior

Memory usage should stabilize after pipeline executions are completed, and unused memory should be reclaimed properly.

### System Info

```shell
Woodpecker Version: 2.7.0
Kubernetes Version: v1.29.4
Environment: Running Woodpecker on a Kubernetes cluster
Number of agents: 10
```

### Additional context

I am using golang profiling to find something about it, this is what I could find so far: 
<img width="1270" alt="Captura de Tela 2024-10-14 às 14 03 23" src="https://github.com/user-attachments/assets/5e012749-a2a0-4bc5-ba52-5e5b5691599c">
![woodpeckergraph](https://github.com/user-attachments/assets/0f42808f-09f3-44da-92d8-87796f39091a)

Has anyone ever faced an issue like this?

### Validations

- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]

**Comment by zc-devs** [comment]
> Has anyone ever faced an issue like this?

Not me. But have no such a load (10 agents) :)

1. When did it start / what is the behavior on the previous versions? Have you tested on `2.7.1`, `next`?
2. How to gather this `pprof` statistic? Is there some guide? I didn't find anything in the WP docs.
3. Nice `pprof` info, but this screens from an agent, which allocated 44.36 MB of memory, if I understand correctly. However, Grafana shows memory usage around 1 GB and that is the issue (I suppose). It would be nice if you had `pprof` stats from a mentioned agent.
4. What is the load? I mean `WOODPECKER_MAX_WORKFLOWS` and how many do you run simultaneously?
Could you explain the right half of the Grafana chart? Something like:
- at this point we run 1 pipeline with 10 workflows
- at this point they all finished
- at this point we run another 10 pipelines with 1 workflow
- at this point they finished and there were no load at all for next 1 hour
5. What is the config of the Server? How much instances? What's about database? What is the load on Server and database?
6. Where do you store the pipeline (steps) logs?

**Comment by lara-clink** [comment]
Hey @zc-devs , we are currently working on our migration project (automated migration from Drone CI to Woodpecker) and I could not collect all of the answers for you yet. By the end of this week I should be able to come back to that.

**Comment by lara-clink** [comment]
> > Has anyone ever faced an issue like this?
> 
> Not me. But have no such a load (10 agents) :)
> 
> 1. When did it start / what is the behavior on the previous versions? Have you tested on `2.7.1`, `next`?
> 2. How to gather this `pprof` statistic? Is there some guide? I didn't find anything in the WP docs.
> 3. Nice `pprof` info, but this screens from an agent, which allocated 44.36 MB of memory, if I understand correctly. However, Grafana shows memory usage around 1 GB and that is the issue (I suppose). It would be nice if you had `pprof` stats from a mentioned agent.
> 4. What is the load? I mean `WOODPECKER_MAX_WORKFLOWS` and how many do you run simultaneously?
>    Could you explain the right half of the Grafana chart? Something like:
> 
> * at this point we run 1 pipeline with 10 workflows
> * at this point they all finished
> * at this point we run another 10 pipelines with 1 workflow
> * at this point they finished and there were no load at all for next 1 hour
> 
> 5. What is the config of the Server? How much instances? What's about database? What is the load on Server and database?
> 6. Where do you store the pipeline (steps) logs?

1. We started to use woodpecker in 2.3.0, and since that we are facing memory leak issues, so we can not know since which version the problem occurs. We have not tested later versions since 2.7.0;
2. We ran a forked version from 2.7.0
I used this tutorial to do it: https://hackernoon.com/go-the-complete-guide-to-profiling-your-code-h51r3waz;
3. There you go:
<img width="900" alt="Captura de Tela 2024-10-29 às 10 29 21" src="https://github.com/user-attachments/assets/efec88f8-7c39-4c77-ab1d-8132e4baf885">
<img width="1190" alt="Captura de Tela 2024-10-29 às 10 30 41" src="https://github.com/user-attachments/assets/eec9463e-db93-4155-9af2-c015e696e2ec">

4. The WOODPECKER_MAX_WORKFLOWS is 10 and we have 15 pods, so it is 150 workflows simultaneously. But the grafana just shows that memory usage increases as we still use Woodpecker. The low points means just that we had a deployment and the pods restarted; 
5. 
memory: 4Gi
requests:
cpu: '2'
memory: 4Gi

**Comment by zc-devs** [comment]
1. Thank you for the guide. Sadly, it's not so convenient to patch and build own version. Could you make a PR with pprof functionality? It should be optional, so flag like `WOODPECKER_PPROF_ENABLED: true|false`. It would be helpful in the future for all users.
2. What are the [versions of](https://github.com/woodpecker-ci/woodpecker/blob/main/go.mod) 
```
k8s.io/api
k8s.io/apimachinery
k8s.io/client-go
```
in your fork? Have you tried to update it?

**Comment by lara-clink** [comment]
those are: 
	k8s.io/api v0.30.2
	k8s.io/apimachinery v0.30.2
	k8s.io/client-go v0.30.2
	
and we have not tried updating it yet

**Comment by lara-clink** [comment]
Hey @zc-devs , we did some tests updating the libs, first we did on our woodpecker-beta environment and now on production. We got similar results on both tests:
This is the memory usage graph before updating the 3 libraries for v0.31.2 version:
![Captura de Tela 2024-11-18 às 16 07 42](https://github.com/user-attachments/assets/06cc8564-6b62-4e37-bd46-189b7969249c)
And this is after the update:
![Captura de Tela 2024-11-21 às 09 12 33](https://github.com/user-attachments/assets/c70db25f-ac48-4a9c-81e2-4301674d60ba)
As you can see, we are still having memory leak behavior since the pod allocates memory and never release it all. But we had some changes in our profiling results:
<img width="946" alt="Captura de Tela 2024-11-21 às 09 20 04" src="https://github.com/user-attachments/assets/134ec1ab-eadc-407e-af96-e0c10b72efed">
Now only k8s.io/apimachinery shows when we run the "top" command, here is the graph we get when we run the "web" command: 
![pprofwoodpecker](https://github.com/user-attachments/assets/8139936b-856a-4bdd-a346-ff1a254c46c3)
In conclusion, we think that the issue is in the k8s.io/apimachinery library.

**Comment by zc-devs** [comment]
> we are still having memory leak behavior since the pod allocates memory and never releases it

![388505256-c70db25f-ac48-4a9c-81e2-4301674d60ba](https://github.com/user-attachments/assets/24d194b1-17ff-4511-9e59-7a0a428df25b)

Shared informers use cache. So, 10 MB could be Woodpecker Agent itself, and 30 MB could be the caches filling at first pipeline run (10+30=40).

> Showing nodes accounting for 25 MB, 90% of 28 MB total

Do I miss something or you are trying to measure the leak at the start of the Pod? Perhaps, I do not understand how `pprof` works, then correct me, please.

> Nice pprof info, but this screens from an agent, which allocated 44.36 MB of memory, if I understand correctly. However, Grafana shows memory usage around 1 GB and that is the issue (I suppose). It would be nice if you had pprof stats from a mentioned agent.

^ is still valid. Could you get `pprof` info when Agent takes GIGAbytes of memory:
![376330232-0c4fdc5c-bb6b-498f-855a-bd81a7d116eb](https://github.com/user-attachments/assets/293df1e3-b15b-4618-94e2-6ad7e13f13bf)

**Comment by lara-clink** [comment]
The graph does not have any relation with pprof, the point I am making with those graphs is that the memory is never fully released, as you can see in the first graph in your comment, the second release does not reach the same point as the first, it is always a little bit higher.
